File: packages/lumexio/native-line-chart/resources/ios/LineChartRenderer.swift
import SwiftUI

/// Draws the two null-padded series ("historical" solid + filled, "forecast"
/// dashed + filled) as smooth curves, mirroring the web app's Chart.js
/// config (tension 0.35, ~15%/12% fill opacity) with hand-rolled SwiftUI
/// Canvas drawing — no charting library dependency needed for one chart.
struct LineChartRenderer: View {
    let node: NativeUINode

    private struct SeriesData {
        let historical: [Float?]
        let forecast: [Float?]
    }

    private func parseSeries(_ raw: String) -> SeriesData {
        guard let data = raw.data(using: .utf8),
              let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            return SeriesData(historical: [], forecast: [])
        }

        func floats(_ key: String) -> [Float?] {
            guard let array = json[key] as? [Any?] else { return [] }
            return array.map { value -> Float? in
                guard let value, !(value is NSNull), let number = value as? NSNumber else { return nil }
                return number.floatValue
            }
        }

        return SeriesData(historical: floats("historical"), forecast: floats("forecast"))
    }

    /// Catmull-Rom-style smoothing so the curve reads like Chart.js's tension:0.35.
    private func smoothPath(_ points: [CGPoint]) -> Path {
        var path = Path()
        guard let first = points.first else { return path }
        path.move(to: first)
        guard points.count > 1 else { return path }

        let smoothing: CGFloat = 0.2
        for i in 0..<(points.count - 1) {
            let p0 = points[i == 0 ? i : i - 1]
            let p1 = points[i]
            let p2 = points[i + 1]
            let p3 = points[i + 2 < points.count ? i + 2 : i + 1]

            let c1 = CGPoint(x: p1.x + (p2.x - p0.x) * smoothing, y: p1.y + (p2.y - p0.y) * smoothing)
            let c2 = CGPoint(x: p2.x - (p3.x - p1.x) * smoothing, y: p2.y - (p3.y - p1.y) * smoothing)

            path.addCurve(to: p2, control1: c1, control2: c2)
        }

        return path
    }

    var body: some View {
        let p = node.props
        let series = parseSeries(p.getString("data", default: ""))
        let historicalColor = Color(argb: p.getColor("historical_color", default: 0xFF2D5D5A))
        let forecastColor = Color(argb: p.getColor("forecast_color", default: 0xFFEC7C0E))
        let selectedIndex = p.getInt("selected_index", default: -1)
        let onChangeCb = p.getCallbackId("on_change")
        let nodeId = node.id
        let count = max(series.historical.count, series.forecast.count)

        // A `let`-bound closure, not a nested `func` — a local function
        // declaration inside `body` fails to compile ("closure containing a
        // declaration cannot be used with result builder 'ViewBuilder'"),
        // since `@ViewBuilder` bodies only tolerate `let`/`var` statements.
        let value: (Int) -> Float? = { index in
            (index < series.historical.count ? series.historical[index] : nil)
                ?? (index < series.forecast.count ? series.forecast[index] : nil)
        }

        GeometryReader { geo in
            Canvas { context, size in
                guard count > 1 else { return }

                let combined = (0..<count).compactMap { value($0) }
                let maxVal = max(combined.max() ?? 1, 1)
                let stepX = size.width / CGFloat(count - 1)

                func point(_ index: Int, _ value: Float) -> CGPoint {
                    let x = CGFloat(index) * stepX
                    let y = size.height - (CGFloat(value) / CGFloat(maxVal)) * size.height
                    return CGPoint(x: x, y: y)
                }

                func draw(_ values: [Float?], color: Color, dashed: Bool) {
                    var segments: [[CGPoint]] = []
                    var current: [CGPoint] = []

                    for (i, v) in values.enumerated() {
                        if let v {
                            current.append(point(i, v))
                        } else if !current.isEmpty {
                            segments.append(current)
                            current = []
                        }
                    }
                    if !current.isEmpty { segments.append(current) }

                    for points in segments {
                        guard points.count > 1, let last = points.last, let first = points.first else { continue }

                        let linePath = smoothPath(points)

                        var fillPath = linePath
                        fillPath.addLine(to: CGPoint(x: last.x, y: size.height))
                        fillPath.addLine(to: CGPoint(x: first.x, y: size.height))
                        fillPath.closeSubpath()

                        context.fill(fillPath, with: .color(color.opacity(0.15)))

                        var strokeStyle = StrokeStyle(lineWidth: 2.5, lineCap: .round, lineJoin: .round)
                        if dashed { strokeStyle.dash = [6, 4] }
                        context.stroke(linePath, with: .color(color), style: strokeStyle)
                    }
                }

                draw(series.historical, color: historicalColor, dashed: false)
                draw(series.forecast, color: forecastColor, dashed: true)

                if selectedIndex >= 0, selectedIndex < count, let v = value(selectedIndex) {
                    let markerPoint = point(selectedIndex, v)
                    let markerColor = selectedIndex < series.historical.count && series.historical[selectedIndex] != nil
                        ? historicalColor
                        : forecastColor

                    var linePath = Path()
                    linePath.move(to: CGPoint(x: markerPoint.x, y: 0))
                    linePath.addLine(to: CGPoint(x: markerPoint.x, y: size.height))
                    context.stroke(linePath, with: .color(markerColor.opacity(0.4)), lineWidth: 1)

                    context.fill(Path(ellipseIn: CGRect(x: markerPoint.x - 4, y: markerPoint.y - 4, width: 8, height: 8)), with: .color(markerColor))
                    context.fill(Path(ellipseIn: CGRect(x: markerPoint.x - 1.5, y: markerPoint.y - 1.5, width: 3, height: 3)), with: .color(.white))
                }
            }
            .contentShape(Rectangle())
            .gesture(
                DragGesture(minimumDistance: 0)
                    .onEnded { drag in
                        guard count > 1, onChangeCb != 0 else { return }
                        let step = geo.size.width / CGFloat(count - 1)
                        let index = Int((drag.location.x / step).rounded())
                        let clamped = min(max(index, 0), count - 1)
                        NativeUIBridge.sendTabChangeEvent(onChangeCb, nodeId: nodeId, index: clamped)
                    }
            )
        }
    }
}

File: packages/lumexio/native-line-chart/resources/android/LineChartRenderer.kt
package com.lumexio.plugins.linechart.ui

import androidx.compose.foundation.Canvas
import androidx.compose.foundation.gestures.detectTapGestures
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.Path
import androidx.compose.ui.graphics.PathEffect
import androidx.compose.ui.graphics.StrokeCap
import androidx.compose.ui.graphics.StrokeJoin
import androidx.compose.ui.graphics.drawscope.Fill
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.input.pointer.pointerInput
import com.nativephp.mobile.ui.nativerender.NativeUIBridge
import com.nativephp.mobile.ui.nativerender.NativeUINode
import org.json.JSONArray
import org.json.JSONObject
import kotlin.math.roundToInt

/**
 * Draws the two null-padded series ("historical" solid + filled, "forecast"
 * dashed + filled) as smooth curves, mirroring the web app's Chart.js
 * config (tension 0.35, ~15%/12% fill opacity) with hand-rolled Compose
 * Canvas drawing — no charting library dependency needed for one chart.
 */
object LineChartRenderer {

    private data class Series(val historical: List<Float?>, val forecast: List<Float?>)

    private fun parseSeries(raw: String): Series {
        if (raw.isBlank()) return Series(emptyList(), emptyList())

        return try {
            val json = JSONObject(raw)
            Series(
                historical = parseFloatArray(json.optJSONArray("historical")),
                forecast = parseFloatArray(json.optJSONArray("forecast"))
            )
        } catch (_: Exception) {
            Series(emptyList(), emptyList())
        }
    }

    private fun parseFloatArray(array: JSONArray?): List<Float?> {
        if (array == null) return emptyList()

        return (0 until array.length()).map { i ->
            if (array.isNull(i)) null else array.optDouble(i).toFloat()
        }
    }

    /** Catmull-Rom-style smoothing so the curve reads like Chart.js's tension:0.35. */
    private fun smoothPath(points: List<Offset>): Path {
        val path = Path()
        if (points.isEmpty()) return path

        path.moveTo(points[0].x, points[0].y)
        if (points.size == 1) return path

        val smoothing = 0.2f
        for (i in 0 until points.size - 1) {
            val p0 = points[if (i == 0) i else i - 1]
            val p1 = points[i]
            val p2 = points[i + 1]
            val p3 = points[if (i + 2 < points.size) i + 2 else i + 1]

            val c1 = Offset(p1.x + (p2.x - p0.x) * smoothing, p1.y + (p2.y - p0.y) * smoothing)
            val c2 = Offset(p2.x - (p3.x - p1.x) * smoothing, p2.y - (p3.y - p1.y) * smoothing)

            path.cubicTo(c1.x, c1.y, c2.x, c2.y, p2.x, p2.y)
        }

        return path
    }

    @Composable
    fun Render(node: NativeUINode, modifier: Modifier) {
        val p = node.props
        val series = parseSeries(p.getString("data", ""))
        val historicalColor = Color(p.getColor("historical_color", 0xFF2D5D5A.toInt()))
        val forecastColor = Color(p.getColor("forecast_color", 0xFFEC7C0E.toInt()))
        val selectedIndex = p.getInt("selected_index", -1)
        val onChangeCb = p.getCallbackId("on_change")
        val nodeId = node.id
        val count = maxOf(series.historical.size, series.forecast.size)

        Canvas(
            modifier = modifier.pointerInput(count) {
                if (count > 1 && onChangeCb != 0) {
                    detectTapGestures { offset ->
                        val step = size.width.toFloat() / (count - 1)
                        val index = (offset.x / step).roundToInt().coerceIn(0, count - 1)
                        NativeUIBridge.sendTabChangeEvent(onChangeCb, nodeId, index)
                    }
                }
            }
        ) {
            if (count < 2) return@Canvas

            fun valueAt(index: Int): Float? = series.historical.getOrNull(index) ?: series.forecast.getOrNull(index)

            val maxVal = (0 until count).mapNotNull { valueAt(it) }.maxOrNull()?.takeIf { it > 0f } ?: 1f
            val stepX = size.width / (count - 1)

            fun toOffset(index: Int, value: Float): Offset {
                val x = index * stepX
                val y = size.height - (value / maxVal) * size.height
                return Offset(x, y)
            }

            fun drawSeries(values: List<Float?>, color: Color, dashed: Boolean) {
                val segments = mutableListOf<List<Offset>>()
                var current = mutableListOf<Offset>()

                values.forEachIndexed { i, v ->
                    if (v != null) {
                        current.add(toOffset(i, v))
                    } else if (current.isNotEmpty()) {
                        segments.add(current)
                        current = mutableListOf()
                    }
                }
                if (current.isNotEmpty()) segments.add(current)

                segments.forEach { points ->
                    if (points.size < 2) return@forEach

                    val linePath = smoothPath(points)

                    val fillPath = Path().apply {
                        addPath(linePath)
                        lineTo(points.last().x, size.height)
                        lineTo(points.first().x, size.height)
                        close()
                    }
                    drawPath(fillPath, color = color.copy(alpha = 0.15f), style = Fill)

                    drawPath(
                        path = linePath,
                        color = color,
                        style = Stroke(
                            width = 5f,
                            cap = StrokeCap.Round,
                            join = StrokeJoin.Round,
                            pathEffect = if (dashed) PathEffect.dashPathEffect(floatArrayOf(18f, 12f)) else null
                        )
                    )
                }
            }

            drawSeries(series.historical, historicalColor, dashed = false)
            drawSeries(series.forecast, forecastColor, dashed = true)

            if (selectedIndex in 0 until count) {
                valueAt(selectedIndex)?.let { v ->
                    val point = toOffset(selectedIndex, v)
                    val markerColor = if (series.historical.getOrNull(selectedIndex) != null) historicalColor else forecastColor

                    drawLine(
                        color = markerColor.copy(alpha = 0.4f),
                        start = Offset(point.x, 0f),
                        end = Offset(point.x, size.height),
                        strokeWidth = 2f
                    )
                    drawCircle(color = markerColor, radius = 7f, center = point)
                    drawCircle(color = Color.White, radius = 3f, center = point)
                }
            }
        }
    }
}
